src/util/opaque.h

```diff
@@ -0,0 +1,197 @@
+#pragma once
+
+#include <cstdlib>
+#include <memory>
+#include <optional>
+#include <utility>
+#include <variant>
+
+namespace bpftrace::util {
+
+// OpaqueValue corresponds to a blob of type-erased memory. It owns its own
+// underlying memory, taking a copy if necessary. This allows it to be used
+// to read maps, rings, and other locations which may not be permanent.
+//
+// This class does not really provide safety beyond first-order memory safety.
+class OpaqueValue {
+private:
+  struct OwnedBuffer {
+    OwnedBuffer(char *data) : data(data) {};
+    ~OwnedBuffer()
+    {
+      free(data);
+    }
+    char *data;
+  };
+  using SharedMemory = std::shared_ptr<OwnedBuffer>;
+  using Variant = std::variant<uintptr_t, SharedMemory>;
+
+public:
+  // Allows for the creation of an OpaqueValue.
+  static OpaqueValue alloc(size_t length, std::function<void(char *)> init)
+  {
+    return { length, init };
+  }
+
+  // Create a copy of a string.
+  static OpaqueValue string(const std::string &s, size_t sz)
+  {
+    return { sz, [&](char *data) {
+              size_t s_sz = std::min(s.size() + 1, sz);
+              memcpy(data, s.data(), s_sz);
+              memset(data + s_sz, 0, sz - s_sz);
+            } };
+  }
+
+  // Merge two OpaqueValues together.
+  OpaqueValue operator|(const OpaqueValue &other) const
```

Comment:
Does this operator need to be public? I'd prefer hiding as much complexity (templates, overloaded operators, etc.) as possible inside the classes.

Reply:
I've replaced it with `operator+` which is more standardized, I'm good with that.